core/range_scan_load_balancer.cxx

```diff
@@ -0,0 +1,145 @@
+/* -*- Mode: C++; tab-width: 4; c-basic-offset: 4; indent-tabs-mode: nil -*- */
+/*
+ *   Copyright 2024. Couchbase, Inc.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+
+#include "range_scan_load_balancer.hxx"
+
+#include <algorithm>
+#include <limits>
+#include <map>
+#include <numeric>
+#include <queue>
+#include <random>
+#include <vector>
+
+namespace couchbase::core
+{
+range_scan_node_state::range_scan_node_state(std::queue<std::uint16_t> vbuckets)
+  : pending_vbuckets_{ std::move(vbuckets) }
+{
+}
+
+auto
+range_scan_node_state::fetch_vbucket_id() -> std::optional<std::uint16_t>
+{
+    std::lock_guard<std::mutex> const lock{ mutex_ };
+    if (pending_vbuckets_.empty()) {
+        return {};
+    }
+    active_stream_count_++;
+    auto vbucket_id = pending_vbuckets_.front();
+    pending_vbuckets_.pop();
+    return vbucket_id;
+}
+
+void
+range_scan_node_state::notify_stream_ended()
+{
+    std::lock_guard<std::mutex> const lock{ mutex_ };
+    active_stream_count_--;
+}
+
+void
+range_scan_node_state::enqueue_vbucket(std::uint16_t vbucket_id)
+{
+    std::lock_guard<std::mutex> const lock{ mutex_ };
+    pending_vbuckets_.push(vbucket_id);
+}
+
+auto
+range_scan_node_state::active_stream_count() -> std::uint16_t
+{
+    std::lock_guard<std::mutex> const lock{ mutex_ };
+    return active_stream_count_;
+}
+
+auto
+range_scan_node_state::pending_vbucket_count() -> std::size_t
+{
+    std::lock_guard<std::mutex> const lock{ mutex_ };
+    return pending_vbuckets_.size();
+}
+
+range_scan_load_balancer::range_scan_load_balancer(const topology::configuration::vbucket_map& vbucket_map,
+                                                   std::optional<std::uint64_t> seed)
+  : seed_{ seed }
+{
+    std::map<std::int16_t, std::queue<std::uint16_t>> node_to_vbucket_map{};
+    for (std::uint16_t vbucket_id = 0; vbucket_id < vbucket_map.size(); vbucket_id++) {
+        auto node_id = vbucket_map[vbucket_id][0];
+        if (node_to_vbucket_map.count(node_id) == 0) {
+            node_to_vbucket_map[node_id] = {};
+        }
```

Comment:
no need to initialize the queue, it is default-constructible, and `map::operator[]` is non-const, it access-or-insert operator.

godbolt: https://godbolt.org/z/Eo3b8G89f
reference: https://en.cppreference.com/w/cpp/container/map/operator_at
> Returns a reference to the value that is mapped to a key equivalent to key or x respectively, performing an insertion if such key does not already exist.

Reply:
Thanks for the tip, done